### ISSUE TITLE
Single-cell data results are inconsistently reproduced

### ISSUE THREAD

**Issue by xfchen0912** [theislab/PathReg#1]
Hi,
Very nice work, but I'm not getting consistent results when running `notebooks/single cell/single_cell_PathReg.ipynb`
The example results are very different from what I get, can you please give me a solution?

<img width="761" alt="image" src="https://github.com/theislab/PathReg/assets/92902438/d70f5867-d048-4ef6-b3d6-dec78d40e5d0">

![download](https://github.com/theislab/PathReg/assets/92902438/dfdca67a-b9b1-4450-a28a-5e1660e2c1cb)

thanks

**Comment by lila167** [comment]
This is strange. How does your loss look like?

**Comment by xfchen0912** [comment]
This is my loss, 
![image](https://github.com/theislab/PathReg/assets/92902438/19a6f4b4-a672-412b-bac7-f7cc4a93db2d)
![image](https://github.com/theislab/PathReg/assets/92902438/58c47683-77bf-4efe-ac5f-f80c2d2d2492)
More detail code, outputs and model weights, I've emailed to you, please check it out!

**Comment by lila167** [comment]
I can't debug your code but can help you to find the issue. 
It seems the losses are very similar. It might be how you constructed the adjacency matrix. Are you following the same code?

**Comment by xfchen0912** [comment]
Thank you for your reply, but I did not change the code in the jupyter notebook, especially the part of `Adjacency Matrix Analysis`
So could you upload a copy of the checkpoint(single_cell_0.01.pth) used in the example or check the code in the Adjacency Matrix Analysis section.

**Comment by xfchen0912** [comment]
The method of constructing the adjacency matrix in jupyter notebook is as follows:
```
for i, layer in enumerate(model[1].odefunc.layers):
    if i ==0:
        WM = torch.abs(layer.sample_weights_ones())
    else:
        WM = torch.matmul(WM,torch.abs(layer.sample_weights_ones()))
WM = WM.T.cpu().detach().numpy()
WM[np.abs(WM)<1e-5] = 0.

df_WM = pd.DataFrame(WM,index=list(genes_12),columns = list(genes_12))
```
I run it step by step, the outputs are:
<img width="969" alt="image" src="https://github.com/theislab/PathReg/assets/92902438/e72f511c-6032-4c13-8364-5ead56e8fcc1">

I also think there is a problem with the steps to construct the adjacency matrix, could you fix it please?

**Comment by lila167** [comment]
Due to a severe server disruption in my previous institute and my move, I cannot access my data right now. I'll rerun the code to check. In the meanwhile, could you please be more clear what you mean by a problem: "I also think there is a problem with the steps to construct the adjacency matrix, could you fix it please?" What should I fix? It's a notebook.

**Comment by xfchen0912** [comment]
Thanks for your help, what I mean is that running the same code and example data did not get the expected sparse regulatory network and I don't have more detailed information to figure out if it's a problem with the model training part or the construction of the adjacency matrix

It may be that the version inconsistency of some packages causes changes in the calculation method of the some functions, or there is a problem in some parts of the notebook. 

so I hope you can run it again to check it out and provide your virtual environment configuration file(.yml)

**Comment by lila167** [comment]
I think I found the issue. The adjacency matrix should be measured using the following code:
```python 
for i, layer in enumerate(model[1].odefunc.layers):
    if i ==0:
        WM = torch.abs(layer.sample_weights())
    else:
        WM = torch.matmul(WM,torch.abs(layer.sample_weights()))
WM = WM.T.cpu().detach().numpy()
WM[np.abs(WM)<1e-5] = 0.

df_WM = pd.DataFrame(WM,index=list(genes_12),columns = list(genes_12))
```

The only thing I changed is using 'layer.sample_weights()' instead of 'layer.sample_weights_ones()'. Sorry for the mistake and thanks for raising it.

**Comment by xfchen0912** [comment]
Thank you very much for your help and I'll close the issue.
